server/src/unifyfs_p2p_rpc.c

```diff
@@ -315,35 +316,37 @@ static void find_extents_rpc(hg_handle_t handle)
         margo_free_input(handle, &in);
     }
 
-    /* fill rpc response struct with output values */
-    hg_bulk_t bulk_resp_handle;
-    find_extents_out_t out;
-    out.ret = ret;
-    out.num_locations = 0;
+    /* define a bulk handle to transfer chunk address info */
+    hg_bulk_t bulk_resp_handle = HG_BULK_NULL;
     if (ret == UNIFYFS_SUCCESS) {
-        void* buf = (void*) chunk_locs;
-        size_t buf_sz = (size_t)num_chunks * sizeof(chunk_read_req_t);
-        hret = margo_bulk_create(mid, 1, &buf, &buf_sz,
-                                 HG_BULK_READ_ONLY, &bulk_resp_handle);
-        if (hret != HG_SUCCESS) {
-            LOGERR("margo_bulk_create() failed");
-            ret = UNIFYFS_ERROR_MARGO;
-        } else {
-            out.num_locations = num_chunks;
-            out.locations = bulk_resp_handle;
+        if (num_chunks > 0) {
+            void* buf = (void*) chunk_locs;
+            size_t buf_sz = (size_t)num_chunks * sizeof(chunk_read_req_t);
+            hret = margo_bulk_create(mid, 1, &buf, &buf_sz,
+                                     HG_BULK_READ_ONLY, &bulk_resp_handle);
+            if (hret != HG_SUCCESS) {
+                LOGERR("margo_bulk_create() failed");
+                ret = UNIFYFS_ERROR_MARGO;
+            }
         }
     }
 
+    /* fill rpc response struct with output values */
+    find_extents_out_t out;
+    out.ret           = ret;
+    out.num_locations = num_chunks;
+    out.locations     = bulk_resp_handle;
+
     /* send output back to caller */
     hret = margo_respond(handle, &out);
     if (hret != HG_SUCCESS) {
         LOGERR("margo_respond() failed");
     }
-    if (out.num_locations) {
-        margo_bulk_free(bulk_resp_handle);
-    }
 
     /* free margo resources */
+    if (bulk_resp_handle != HG_BULK_NULL) {
+        margo_bulk_free(bulk_resp_handle);
+    }
```

Comment:
I'm not sure if it is possible for the above ``margo_bulk_create`` to fail and set ``hret`` to something other than ``HG_SUCCESS`` but still fill in the ``bulk_resp_handle`` handle value with something other than ``HG_BULK_NULL``.  And if that is possible, do we need to call ``margo_bulk_free`` or not?  I don't know.

Reply:
I would need to check the margo code, but I assume they do the right thing and don't assign to the output handle parameter when the operation fails. The only safe thing we can do is to only call `margo_bulk_free()` when `margo_bulk_create()` was successful. That means you should use the return value from the call to the create as the if condition.

Reply:
I checked the margo code, it just passes straight through to mercury hg_bulk_create(), which only sets the output parameter on success.

Reply:
https://github.com/mercury-hpc/mercury/blob/v1.0.1/src/mercury_bulk.c

Reply:
Cool.  Thanks, @MichaelBrim